### PR TITLE
Fix VB Else statement parsing

### LINKED ISSUE #47109: VB.NET - Else with Expression doesn't throw Compiler Error
The following code doesn't throw a compiler error.
```VB.NET
Dim str = "Foo"
If str.StartsWith("bar") Then
    Console.WriteLine("bar")
Else str.StartsWith("baz") 
    Console.WriteLine("baz")
End If
```
Hint, the `Else` should be an `Else If`

**Version Used**: 
Microsoft Visual Studio Enterprise 2019
Version 16.7.1

**Expected Behavior**:
A compile error (missing if keyword) at the `Else` line

**Actual Behavior**:
No error. The output on console is *baz*

### PR DESCRIPTION
**Pull request by Youssef1313**
Fixes #47109

### CODE REVIEW

File: src/Compilers/VisualBasic/Portable/Parser/ParseStatement.vb [hunk 1]
```diff
@@ -434,6 +434,9 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
             GetNextToken()
 
             Dim statement = SyntaxFactory.ElseStatement(elseKeyword)
+            If Not Context.IsLineIf AndAlso Not CurrentToken.IsEndOfLine Then
+                ReportSyntaxError(elseKeyword, ERRID.ERR_ExpectedEOS)
```
Comment (Youssef1313) [inline src/Compilers/VisualBasic/Portable/Parser/ParseStatement.vb:438]: This line is hit, but I can't see any diagnostics. I don't know what could be the issue.

File: src/Compilers/VisualBasic/Portable/Parser/ParseStatement.vb [hunk 2]
```diff
@@ -434,6 +434,11 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
             GetNextToken()
 
             Dim statement = SyntaxFactory.ElseStatement(elseKeyword)
+            Dim isElseFollowedByNewLineOrColon =
+                (CurrentToken.Kind = SyntaxKind.EndOfLineTrivia OrElse CurrentToken.Kind = SyntaxKind.ColonTrivia)
+            If Not Context.IsLineIf AndAlso Not isElseFollowedByNewLineOrColon Then
+                statement = ReportSyntaxError(statement, ERRID.ERR_ExpectedEOS)
+            End If
```
Comment (Youssef1313) [inline src/Compilers/VisualBasic/Portable/Parser/ParseStatement.vb:441]: Running Visual Studio with my changes produces assertion failure :-(

image

**Comment by 333fred** [comment]
@Youssef1313 I'm going to mark this as a draft until you're ready for review.

**Comment by Youssef1313** [comment]
@333fred I had stopped working on this PR for 2 reasons:

1. Based on the issue comments, I'm not sure if you'll want to make this breaking change.
2. I wasn't able to determine the reason for the assertion failure 😄 

If you'll accept this breaking change, I'll try to give it another try and see what I'll get.

**Comment by 333fred** [comment]
@AlekseyTs, do you have thoughts on this?

**Comment by AlekseyTs** [comment]
If native compiler doesn't report an error, we probably don't want to make a breaking change

**Comment by Youssef1313** [comment]
Native compiler reports an error

**Comment by AlekseyTs** [comment]
If that is the case, I will be comfortable fixing the bug. If you summarize the impact of the fix, we can discuss it with Compatibility Council and get final decision on whether we are willing to accept the breaking change.

**Comment by Youssef1313** [comment]
@AlekseyTs The only benefit I can see from this is pointing to potentially missing "If" (i.e. the code was meant to be "Else If" but was only written as "Else". Especially that there is no warning for unused return value.

For example, the following snippet which is taken from the issue.

```vb
Dim str = "Foo"
If str.StartsWith("bar") Then
    Console.WriteLine("bar")
Else str.StartsWith("baz") 
    Console.WriteLine("baz")
End If
```

**Comment by AlekseyTs** [comment]
I meant breaking impact, what scenarios will be broken and when.

**Comment by Youssef1313** [comment]
> 
> 
> I meant breaking impact, what scenarios will be broken and when.

I'm not sure. But I think it will impact a few people only. Never seen statements written in the same line with "Else" before.

**Comment by Youssef1313** [comment]
@AlekseyTs Can you look at this code comment?

https://github.com/dotnet/roslyn/blob/e704ca635bd6de70a0250e34c4567c7a28fa9f6d/src/Compilers/VisualBasic/Portable/Parser/BlockContexts/IfPartContext.vb#L79-L81

This seems to be related. It says that this is not a syntax error, but that explanation doesn't seem to be enough because this should be disallowed.

**Comment by Youssef1313** [comment]
The following test is also related:

https://github.com/dotnet/roslyn/blob/bcc80c0816c233d98821611ee14861b40b39c7c7/src/Compilers/VisualBasic/Test/Syntax/Parser/ParseStatements.vb#L2075-L2090

**Comment by PathogenDavid** [comment]
While I do agree this is a bug, I don't think this was a regression. I still have Visual Studio 2008 kicking around and it accepts the incorrect syntax:

image

(It does try to automatically insert a colon for me, but it works without it.)

Additionally I built+ran the same test program with just about every Microsoft version of `vbc.exe` WizFile found on my machine and they all build fine and product the same output:

image

> See this online compiler https://www.onlinegdb.com/online_vb_compiler which seems to be based on the native compiler.

Read the output of that website a bit closer. It's using the Mono VB compiler:

```
Visual Basic.Net Compiler version 0.0.0.5943 (Mono 4.6 - tarball)
Copyright (C) 2004-2010 Rolf Bjarne Kvinge. All rights reserved.
```

**Comment by PathogenDavid** [comment]
Did Visual Basic get warning waves like C#? Maybe this should become a warning instead.

**Comment by Youssef1313** [comment]
@PathogenDavid Thanks for investigating further and for the clarifications!

Currently, VB didn't get any warning waves, they may come in future.

Given that this bug exists even in VS 2008, I'm not sure if this should be in a future warning wave or a compiler-error.

**Comment by PathogenDavid** [comment]
@Youssef1313 No problem, I [got some amusement out of it](https://twitter.com/PathogenDavid/status/1306181892958556161).

I have no horse in this race since I haven't used Visual Basic professionally for like 10 years or so, but my initial reaction is that at most this could become a warning if VB ever gets warning waves. It may be a bug, but it's a 19 year old bug that people are probably unintentionally relying on.

**Comment by Youssef1313** [comment]
> people are probably unintentionally relying on.

@PathogenDavid A quick search through GitHub, I found this: 😕 

https://github.com/Underwearliu/Chinese-Chess/blob/bdf4a4ea13306e5f309d0a70080669d59295a21d/Chinese%20Chess%20v2/Player.vb#L33-L36

But I don't think people are relying on this very often. Probably for 95% or more, the usage of Else is followed by a statement terminator.

**Comment by 333fred** [comment]
Given that this is not a regression from the native compiler, I think I'd prefer if we left this as is.

**Comment by Youssef1313** [comment]
@333fred Does the spec need to be updated to reflect this?

**Comment by 333fred** [comment]
I'll let you know.

**Comment by AlekseyTs** [comment]
Closing this PR because we decided to "Won't fix" the bug.

**Comment by 333fred** [comment]
We're also going to leave the spec as is. Much like other parts of roslyn, we have a bug that we're leaving in for compatability, but that shouldn't be reflected in the spec itself.

**Pull request closed without merging**